### DISCUSSION TITLE
Zombie processes when using the runInBackground method in the schedule.

### DISCUSSION THREAD

**Discussion by realadilet** [laravel/framework#37175] [Q&A]
- Laravel Version: 8.27.0
- PHP Version: 7.4.13
- Database Driver & Version: PostgreSQL 13.2 

### Description:

If I add runInBackground method in my schedule tasks, a large number of dead processes appear.

### Steps To Reproduce:

```php
$schedule->command('emails:send')
     ->everyMinute();
     ->runInBackground();
```

command `ps fax | grep php` shows something like that 

```
  55796 ?        Z      0:00      \_ [php] <defunct>
  55892 ?        Z      0:00      \_ [php] <defunct>
  56059 ?        Z      0:00      \_ [php] <defunct>
  56141 ?        Z      0:00      \_ [php] <defunct>
  56144 ?        Z      0:00      \_ [php] <defunct>
  56147 ?        Z      0:00      \_ [php] <defunct>
  56150 ?        Z      0:00      \_ [php] <defunct>
  56153 ?        Z      0:00      \_ [php] <defunct>
  56294 ?        Z      0:00      \_ [php] <defunct>
  56382 ?        Z      0:00      \_ [php] <defunct>
  56500 ?        Z      0:00      \_ [php] <defunct>
```

this will continue indefinitely until you restart docker container.

Also, I have some questions about the implementation of the `runInBackground` method.

There is `buildBackgroundCommand` method, which call [ensureCorrectUser](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Console/Scheduling/CommandBuilder.php#L73) method.

```php
    protected function ensureCorrectUser(Event $event, $command)
    {
        return $event->user && ! windows_os() ? 'sudo -u '.$event->user.' -- sh -c \''.$command.'\'' : $command;
    }
```

1) According to my observations, the `$event->user` is always empty. It is not clear what kind of user and I did not find where the user value is assigned.

2) Even if the user is not empty, how correct is it to execute such a command, especially in a docker container, where there is usually no `sudo` command?

Since user is empty, I have such as command: 

```bash
('/usr/local/bin/php' 'artisan' emails:send > '/dev/null' 2>&1 ; '/usr/local/bin/php' 'artisan' schedule:finish "framework/schedule-cc419d041de759098c4de7e4bce7e1771c60c08a" "$?") > '/dev/null' 2>&1 &
```

How can such a command spawn zombie processes ?

**Comment by ohmydevops** [comment]
same problem.
i'm working on it.

**Comment by AlexLyubimov** [comment]
@taylorotwell same problem, laravel 8.37.0

**Comment by ohmydevops** [comment]
Hi, how do you run your schedule tasks in container?
@adiletmaks @AlexLyubimov

**Reply by AlexLyubimov** [reply]
@amirbagh75 yep

**Reply by ohmydevops** [reply]
I mean, which init system did you use and how?

**Reply by realadilet** [reply]
> I mean, which init system did you use and how?

inside docker OS - Debian GNU/Linux 10 (buster)

**Reply by ohmydevops** [reply]
thanks, but which init system spawn all processes ?

**Reply by realadilet** [reply]
> thanks, but which init system spawn all processes ?

I don't understand what you mean, system spawn processes after run schedule command

**Reply by ohmydevops** [reply]
For example, with supervisord?

**Reply by realadilet** [reply]
> For example, with supervisord?

We don't use supervisord, we start the process in docker at the start. like `exec php artisan schedule:run`

**Comment by nakashu** [comment]
Same problem

Running with:

Laravel: `8.47.0`
Inside docker container using OS crontab.

crontab
```
* * * * * docker-compose -f /home/app/docker-compose.yml exec -T app /usr/local/bin/php /var/www/html/app/artisan schedule:run >/root/app.log 2>&1
```

Output of schedule:run commands:
```
[2022-02-08T14:50:01+00:00] Running scheduled command: ('/usr/local/bin/php' 'artisan' page:snap >> '/var/www/html/app/storage/logs/page_snap.log' 2>&1 ; '/usr/local/bin/php' 'artisan' schedule:finish "framework/schedule-18b90dc34521c3c26633b2bfb3a9f4af054fecb5" "$?") > '/dev/null' 2>&1 &
```

No error, everything passes correctly, but zombie processes remain.

**Comment by 675076143** [comment]
Hi, check the pid=1 problem.
https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/

**Comment by maccup** [comment]
run your docker with "--init" flag and this will solve the problem.

**Reply by xx10n31y** [reply]
Is that working on kubelet ?

**Comment by BetaJason** [comment]
I fixed this by including `pcntl_waitpid(-1, $status, WNOHANG);` in the end of the `schedule()` function of `App/Console/Kernel.php`

**Comment by fabriciojs** [comment]
Using a proper init process solved it - in my case [`tini`](https://github.com/krallin/tini) was the chosen one.

**Comment by maoxuner** [comment]
init（pid 0, 比如crond）进程接管了schedule主进程（php artisan schedule:run）遗留的shedule子进程（runInBackground，比如 php artisan command-name），却不主动负责销毁这些子进程。主进程并不会等待并销毁runInBackground的子进程，于是这些子进程就变成了僵尸进程。

schedule主进程不等待并销毁是间接原因，init进程不负责销毁是直接原因。前者的行为是符合预期的，如果后者是crond，那也是符合预期的。但可以把后者替换成具备销毁孤儿进程能力的程序，比如s6 init。然后通过s6启动crond。

在此推荐使用 [s6-overlay](https://github.com/just-containers/s6-overlay)

---

The init process (PID 0, such as crond) assumes control over child processes left behind by the primary scheduling process (php artisan schedule:run), which are designed to run in the background (like php artisan command-name with runInBackground).

However, init does not actively take on the responsibility of terminating these child processes. The main scheduling process does not wait for and destroy these background-running child processes, causing them to become zombie processes.

The lack of waiting and cleanup by the primary scheduling process is an indirect cause, while init's failure to handle their destruction is the direct reason. The former's behavior aligns with expectations; if the latter is indeed crond, this too is expected behavior. Nonetheless, an alternative solution could involve replacing init with a program capable of reaping orphaned processes, such as s6 init. Then start crond with s6.

It is therefore recommended to utilize [s6-overlay](https://github.com/just-containers/s6-overlay) for addressing this issue.
